### DISCUSSION TITLE
`consume_event` flag

### DISCUSSION THREAD

**Discussion by adamfk** [StateSmith/StateSmith#143] [Ideas]
From discussion here: https://github.com/StateSmith/StateSmith/issues/141

In my years of using state machines, I've only needed to use the `consume_event` functionality a few times. It generally isn't needed. Its syntax and behavior should be considered experimental at this point. We might remove it if nobody actually uses it.

Some background: https://github.com/StateSmith/StateSmith/blob/main/docs/diagram-features.md

> When an event is dispatched to a state machine, the active state first gets a chance to handle it. If it doesn't handle the event, then its parent gets a chance... all the way up to the state machine root.
>
> An event `EV1` is handled/consumed if a state has a behavior with trigger `EV1` and a guard that evaluates to true.
>
> There is one exception - the `do` event. The `do` event is special in that state behaviors don't normally consume it. This allows child and parent states to `do` some work all the way up the chain.
>
> If a transition occurs, no other behaviors will be checked for any state (doesn't matter if it is the `do` event).
>
> State's can choose to not consume an event or consume the `do` event modifying a state's `consume_event` flag (see a state's event handler code).

## Specification Unit Test For `DO` Event Handling

![image](https://user-images.githubusercontent.com/274012/225319480-308d330a-048a-4bc0-b778-fe5661c89802.png)

```
Dispatch event DO
===================================================
State TEST1_S1_1: check behavior `do`. Behavior running.
State TEST1_ROOT: check behavior `do`. Behavior running.

Dispatch event EV1
===================================================
State TEST1_S1_1: check behavior `EV1 TransitionTo(TEST1_S2)`. Behavior running.
Exit TEST1_S1_1.
Exit TEST1_S1.
Transition action `` for TEST1_S1_1 to TEST1_S2.
Enter TEST1_S2.

Dispatch event DO
===================================================
State TEST1_S2: check behavior `do / { consume_event = true; }`. Behavior running.
```

## Specification Unit Test For Regular Event Handling
![image](https://user-images.githubusercontent.com/274012/225321709-25d9b38f-774a-4431-8758-b87526700ada.png)

```
Dispatch event EV2
===================================================
State TEST2_ROOT: check behavior `EV2`. Behavior running.

Dispatch event EV1
===================================================
State TEST2_S1_1: check behavior `EV1`. Behavior running.

Dispatch event DO
===================================================
State TEST2_S1_1: check behavior `do TransitionTo(TEST2_S2)`. Behavior running.
Exit TEST2_S1_1.
Exit TEST2_S1.
Transition action `` for TEST2_S1_1 to TEST2_S2.
Enter TEST2_S2.
State TEST2_S2: check behavior `ENTER`. Behavior running.

Dispatch event EV1
===================================================
State TEST2_S2: check behavior `ev1 / { consume_event = false; }`. Behavior running.
State TEST2_ROOT: check behavior `EV1`. Behavior running.

Dispatch event EV2
===================================================
State TEST2_S2: check behavior `ev2 TransitionTo(TEST2_S2)`. Behavior running.
Exit TEST2_S2.
Transition action `` for TEST2_S2 to TEST2_S2.
Enter TEST2_S2.
State TEST2_S2: check behavior `ENTER`. Behavior running.
```

# Use Cases
One use case comes from Object Oriented Programming. When you look at state machines implemented by users in code, you sometimes see this pattern. A child state does an action on event, but then still wants the parent behavior to follow.

```csharp
class SuperState
{
    protected int count;

    public virtual void HandleEvent1()
    {
        Console.WriteLine("Count is " + count);
        // check transition conditions...
    }
}

class ChildState : SuperState 
{
    public override void HandleEvent1()
    {
        count = 0;
        base.HandleEvent1();
    }
}
```

We can achieve something similar with the `consume_event` flag. Here's a simplified excerpt the [laser tag example](https://github.com/StateSmith/StateSmith/blob/main/examples/LaserTagMenu1/LaserTagMenu1/LaserTagMenu1Sm.graphml).

![image](https://user-images.githubusercontent.com/274012/225326545-783d70c5-e95b-466c-8dc6-34655d0131c6.png)

**Comment by iamjuing** [comment]
Consider defining a special keyword similar to what is mentioned in discussion #167 to avoid the action area and a need to know the internals (for example in case the internal variable changes its name)?

E.g. `<pass> EVENT / action1()`

**Reply by adamfk** [reply]
Opened https://github.com/StateSmith/StateSmith/issues/175 to consider.

The stereotype syntax has merit, but I would still want to be able to set it dynamically so that I can support sub state machines / orthogonal states. See https://github.com/StateSmith/StateSmith/issues/141#issuecomment-1512367272

**Reply by iamjuing** [reply]
My vision here is that neither submachine states nor orthogonal regions should not need anything like that.
Sumachines is quite tricky, orhogonal regions shall be easy as they are solid part of the state machine.

**Reply by adamfk** [reply]
I agree. For normal uses cases (submachine, orthogonal), the user shouldn't have to worry about `consume_event`. Initially, submachine and orthogonal states may be automatically implemented "under the hood" using `consume_event` (seamless to user).

My other use case is that in an upcoming project, I want to use a non-StateSmith state machine as a submachine (3rd party menu system state machine). This will require me (the user) to manually implement the submachine integration. I'll need to use `consume_event` for this.

**Reply by iamjuing** [reply]
W/o knowing the details, is that really the use case for a submachine or rather for two cooperating, yet independent state machines that can exchange events?

My point is not to try to be too generic here as it might become to complex. 

Just an idea...
Submachine implementation could go also in other, simpler direction:
Understanding submachine as a C macro ... just to replace recurring parts of one diagram. There could be a $SUBMACHINE element in the same document, basically a composite state defined out of the main diagram.

**Reply by adamfk** [reply]
> W/o knowing the details, is that really the use case for a submachine or rather for two cooperating, yet independent state machines that can exchange events?

It isn't my main use case for a submachine, but I'd like to retain the flexibility to experiment with it.
In the past, I've done the above (independent cooperating) and it works out OK, but adds complexity to the overall design. I'm curious if treating the 3rd party state machine as a submachine would significantly help or hinder.

> My point is not to try to be too generic here as it might become to complex.

That is a fair point. Definitely something to keep in mind.

I'm adding your implementation idea to https://github.com/StateSmith/StateSmith/issues/179
